OpenTelemetry Emitter Extension (#47)
confluentinc/druid: extensions-contrib/opentelemetry-emitter/README.md

> # OpenTracing Emitter

Reviewer: opentracing?

The OpenTelemetry emitter generates OpenTelemetry Spans for queries.

## Configuration

### Enabling

Load the plugin and enable the emitter in `common.runtime.properties`.

Load the plugin:

```
druid.extensions.loadList=[..., "opentelemetry-emitter"]
```

Enable the emitter:

```

> druid.emitter=opentelemetry

Reviewer: since the opentelemetry emitter doesn't support metrics, it's unlikely someone would only enable that emitter. Maybe the example should be more realistic and show how to add it to a config with an existing emitter.

```

> *_Don't forget to remove the line_ `druid.emitter=noop` _to avoid rewriting_.

Reviewer: not sure what we mean by that "rewriting". Maybe we can link to the common emitter configuration docs for further reading?

    ports:
      - "9411:9411"

  otel-collector:
    image: otel/opentelemetry-collector:latest
    command: ["--config=otel-local-config.yaml", "${OTELCOL_ARGS}"]
    volumes:
      - ${PWD}/config.yaml:/otel-local-config.yaml
    ports:
      - "4317:4317"
```

Create `config.yaml` file with configuration for otel-collector:
```
version: "2"
receivers:
receivers:
  otlp:
    protocols:
      grpc:

exporters:
  zipkin:
    endpoint: "http://zipkin-all-in-one:9411/api/v2/spans"
    format: proto

  logging:

processors:
  batch:

service:
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, zipkin]
```
*_How to configure otel-collector you can read [here](https://opentelemetry.io/docs/collector/configuration/)._

Run otel-collector and zipkin.
```
docker-compose up
```

### Part 2: Run druid
Build druid:

```
mvn clean install -Pdist
tar -C /tmp -xf distribution/target/apache-druid-0.21.0-bin.tar.gz
cd /tmp/apache-druid-0.21.0
```

Edit `conf/druid/single-server/micro-quickstart/_common/common.runtime.properties` to enable
the emitter (see `Configuration` section above).

Start the quickstart:

```
bin/start-micro-quickstart
```

> Load sampel data - [example](https://druid.apache.org/docs/latest/tutorials/index.html#step-4-load-data).

Reviewer 2: ```suggestion
Load sample data - [example](https://druid.apache.org/docs/latest/tutorials/index.html#step-4-load-data).
```

### Part 3: Send queries

Create `query.json`:
```
{

>    "query":"YOUR QUERY",

Reviewer: to make this very user-friendly you could even add a query that works with the sample data loaded above

      receivers: [otlp]
      processors: [batch]
      exporters: [logging, zipkin]
```
*_How to configure otel-collector you can read [here](https://opentelemetry.io/docs/collector/configuration/)._

Run otel-collector and zipkin.
```
docker-compose up
```

### Part 2: Run Druid
Build Druid:

```
mvn clean install -Pdist
tar -C /tmp -xf distribution/target/apache-druid-0.21.0-bin.tar.gz
cd /tmp/apache-druid-0.21.0
```

Edit `conf/druid/single-server/micro-quickstart/_common/common.runtime.properties` to enable
the emitter (see `Configuration` section above).

Start the quickstart with the apppropriate environment variables for opentelemetry autoconfiguration:
```
OTEL_SERVICE_NAME="org.apache.druid" OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:4317" bin/start-micro-quickstart
```
*_More about opentelemetry autoconfiguration [here](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure)_

Load sample data - [example](https://druid.apache.org/docs/latest/tutorials/index.html#step-4-load-data).

### Part 3: Send queries

Create `query.json`:
```
{
   "query":"SELECT COUNT(*) as total FROM wiki WHERE countryName IS NOT NULL",
   "context":{

>       "traceparent":"00-54ef39243e3feb12072e0f8a74c1d55a-ad6d5b581d7c29c1-01"

Author: > What happens if the traceparent is not set in the context? Will one be automatically generated internally?
If the traceparent is not set in the context, then spans are send to exporter without traceParentId and spanParentId

# OpenTelemetry Emitter

The [OpenTelemetry](https://opentelemetry.io/) emitter generates OpenTelemetry Spans for queries.

## How OpenTelemetry emitter works

> The [OpenTelemetry](https://opentelemetry.io/) emitter processes`ServiceMetricEvent` type of event with a `query/time`

Reviewer: ```suggestion
The [OpenTelemetry](https://opentelemetry.io/) emitter processes `ServiceMetricEvent` type of event with a `query/time`
```

metric. It extacts OpenTelemetry context

> from [druid context](https://druid.apache.org/docs/latest/querying/query-context.html). To link druid spans to parent

Reviewer: ```suggestion
from the [query context](https://druid.apache.org/docs/latest/querying/query-context.html). To link druid spans to parent
```

trace druid context should contain `traceparent` key, at least. More

> about [context propagation](https://www.w3.org/TR/trace-context/). If there is no `traceparent` key is provided, then

Reviewer: ```suggestion [context propagation](https://www.w3.org/TR/trace-context/) for more information. If no `traceparent` key is provided, then
```

spans are created without `parentTraceId` and link to no parent. Also, the emitter adds other druid context entries to
the span attributes.

## Configuration

### Enabling

> Load the plugin and enable the emitter in `common.runtime.properties`.

Reviewer: ```suggestion
To enable the OpenTelemetry emitter, add the extension and enable the emitter in `common.runtime.properties`.
```

# OpenTracing Emitter

The OpenTelemetry emitter generates OpenTelemetry Spans for queries.

## Configuration

### Enabling

Load the plugin and enable the emitter in `common.runtime.properties`:

Load the plugin:

```
druid.extensions.loadList=[..., "opentelemetry-emitter"]
```

Enable the emitter:

```
druid.emitter=opentelemetry
```

*_Don't forget to remove the line_ `druid.emitter=noop` _to avoid rewriting_.

## Testing
###Zipkin
Run zipkin docker container
```
docker run -d -p 9411:9411 openzipkin/zipkin
```
Zipkin will be available at `http://localhost:9411`.

Add zipkin dependency to `extensions-contrib/opentelemetry-emitter/pom.xml`:
```
<dependency>
  <groupId>io.opentelemetry</groupId>
  <artifactId>opentelemetry-exporter-zipkin</artifactId>
  <version>${your.version}</version>

> </dependency>

Reviewer: this might be fine for development, but requires recompiling. We should aim to make the extension work out of the box. The most straightforward way to do so is probably to bundle the OTLP exporter by default and let users runs their own otel collector or hook up a tracing vendor which supports OTLP out of the box

Author: Yeah, that make sense, I'll remove this section with Zipkin and leave only section with OTLP collector
OTLP exporter is already bundled, so users can use OTLP collector to export traces

```

Run:

```
mvn package
tar -C /tmp -xf distribution/target/apache-druid-0.21.0-bin.tar.gz
cd /tmp/apache-druid-0.21.0
```

Edit `conf/druid/single-server/micro-quickstart/_common/common.runtime.properties` to enable
the emitter (see `Configuration` section above). Set the following properties:
```

> otel.traces.exporter=zipkin

Author: It doesn't work if add it to `common.runtime.properties`
I set this property directly from the code:
`System.setProperty("otel.traces.exporter", "zipkin")`
@xvrl  Could you tell me please if you know how to do it better?

Author: Sure
Usually all settings are stored in config class using jackson annotations, so it's seems good option for our case

otel.service.name=druid
```
[More about OpenTelemetry SDK Autoconfigure](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure)

Start the quickstart:

```
bin/start-micro-quickstart
```

Load sampel data - [example](https://druid.apache.org/docs/latest/tutorials/index.html#step-4-load-data)

Then you can send queries using `curl` and a query file `query.json`:
```
curl -XPOST -H'Content-Type: application/json' http://localhost:8888/druid/v2/sql/ -d @query.json
```

> To test propagation run simple server which read query from stdin:

Reviewer: rather than pasting a bunch of code here, it would be nicer if we could use some existing opentelemetry collector docker images to demo how to test this.

Author: Yeah, absolutely
But we want to put trace context to json and then make request to druid, so I think it's not a common case
Or do you mean to create our own docker image?

Reviewer 3: We can create our own docker image, and imply provide `make` command line in `README.md`, you can checkout this example here in golang: https://github.com/confluentinc/cc-otel-collector/tree/master/example We run the test by `make run-example-client`.
